<a href="https://colab.research.google.com/github/Bageshri123/Bageshri-Portfolio/blob/main/Preparing_data_for_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Skills covered :**
1.  Removing duplicate records from a data set
2.  Replacing null values where necessary



**Exercise  - check for null values**




In the exercise below, I have read the data from a CSV file housing_in_london_yearly_variables.csv. Then used df.isnull().values.any() function to check if there are any null values in th edataframe. It returned 'True'. Then used df.info() to see which columns contain null values. Where the non-null count is less than total entries, it means the column contains missing/NA entries.

In [ ]:
import pandas as pd
url="https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
housing=pd.read_csv(url)

print("Are there any null values in the dataframe?", housing.isnull().values.any(),"\n")

housing.info()

Are there any null values in the dataframe? True 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1071 entries, 0 to 1070
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code               1071 non-null   object 
 1   area               1071 non-null   object 
 2   date               1071 non-null   object 
 3   median_salary      1049 non-null   float64
 4   life_satisfaction  352 non-null    float64
 5   mean_salary        1071 non-null   object 
 6   recycling_pct      860 non-null    object 
 7   population_size    1018 non-null   float64
 8   number_of_jobs     931 non-null    float64
 9   area_size          666 non-null    float64
 10  no_of_houses       666 non-null    float64
 11  borough_flag       1071 non-null   int64  
dtypes: float64(6), int64(1), object(5)
memory usage: 100.5+ KB


**Exercise  -Remove null values**

I have used df.dropna() to remove all NA values from the column life_satisfaction. I have used the parameter 'subset' to specify on which column to use the dropna function. We can see that row count has been reduced to 352 rows.

In [ ]:
#remove rows with NA values for life_satisfaction
housing.dropna(subset=["life_satisfaction"])


code                  area  ... no_of_houses  borough_flag
613   E09000002  barking and dagenham  ...      71079.0             1
614   E09000003                barnet  ...     139346.0             1
615   E09000004                bexley  ...      95037.0             1
616   E09000005                 brent  ...     112083.0             1
617   E09000006               bromley  ...     135036.0             1
...         ...                   ...  ...          ...           ...
1010  E12000009            south west  ...          NaN             0
1013  E92000001               england  ...   24172166.0             0
1017  N92000002      northern ireland  ...          NaN             0
1018  S92000003              scotland  ...          NaN             0
1019  W92000004                 wales  ...          NaN             0

[352 rows x 12 columns]

In the below exercise, I have used dropna() without any parameter to remove null values across the whole dataframe. The row count has now reduced to 267 rows.

In [ ]:
# remove all NA values across whole dataframe
housing.dropna()

code                  area  ... no_of_houses  borough_flag
613   E09000002  barking and dagenham  ...      71079.0             1
614   E09000003                barnet  ...     139346.0             1
615   E09000004                bexley  ...      95037.0             1
616   E09000005                 brent  ...     112083.0             1
617   E09000006               bromley  ...     135036.0             1
...         ...                   ...  ...          ...           ...
999   E09000031        waltham forest  ...     103029.0             1
1000  E09000032            wandsworth  ...     146162.0             1
1001  E09000033           westminster  ...     124509.0             1
1008  E12000007                london  ...    3556161.0             0
1013  E92000001               england  ...   24172166.0             0

[267 rows x 12 columns]

**Exercise  - Removing duplicate entries**

In the below exercise, I have used df.drop_duplicates() with parameter 'subset' to remove duplicate entries for the column 'area'.

In [ ]:
housing.drop_duplicates(subset=['area'])

code                      area  ... no_of_houses  borough_flag
0   E09000001            city of london  ...          NaN             1
1   E09000002      barking and dagenham  ...          NaN             1
2   E09000003                    barnet  ...          NaN             1
3   E09000004                    bexley  ...          NaN             1
4   E09000005                     brent  ...          NaN             1
5   E09000006                   bromley  ...          NaN             1
6   E09000007                    camden  ...          NaN             1
7   E09000008                   croydon  ...          NaN             1
8   E09000009                    ealing  ...          NaN             1
9   E09000010                   enfield  ...          NaN             1
10  E09000011                 greenwich  ...          NaN             1
11  E09000012                   hackney  ...          NaN             1
12  E09000013    hammersmith and fulham  ...          NaN             1
13  E09000014                  haringey  ...          NaN             1
14  E09000015                    harrow  ...          NaN             1
15  E09000016                  havering  ...          NaN             1
16  E09000017                hillingdon  ...          NaN             1
17  E09000018                  hounslow  ...          NaN             1
18  E09000019                 islington  ...          NaN             1
19  E09000020    kensington and chelsea  ...          NaN             1
20  E09000021      kingston upon thames  ...          NaN             1
21  E09000022                   lambeth  ...          NaN             1
22  E09000023                  lewisham  ...          NaN             1
23  E09000024                    merton  ...          NaN             1
24  E09000025                    newham  ...          NaN             1
25  E09000026                 redbridge  ...          NaN             1
26  E09000027      richmond upon thames  ...          NaN             1
27  E09000028                 southwark  ...          NaN             1
28  E09000029                    sutton  ...          NaN             1
29  E09000030             tower hamlets  ...          NaN             1
30  E09000031            waltham forest  ...          NaN             1
31  E09000032                wandsworth  ...          NaN             1
32  E09000033               westminster  ...          NaN             1
33  E12000001                north east  ...          NaN             0
34  E12000002                north west  ...          NaN             0
35  E12000003  yorkshire and the humber  ...          NaN             0
36  E12000004             east midlands  ...          NaN             0
37  E12000005             west midlands  ...          NaN             0
38  E12000006                      east  ...          NaN             0
39  E12000007                    london  ...          NaN             0
40  E12000008                south east  ...          NaN             0
41  E12000009                south west  ...          NaN             0
42  E13000001              inner london  ...          NaN             0
43  E13000002              outer london  ...          NaN             0
44  E92000001                   england  ...          NaN             0
45  K02000001            united kingdom  ...          NaN             0
46  K03000001             great britain  ...          NaN             0
47  K04000001         england and wales  ...          NaN             0
48  N92000002          northern ireland  ...          NaN             0
49  S92000003                  scotland  ...          NaN             0
50  W92000004                     wales  ...          NaN             0

[51 rows x 12 columns]

**Skills Covered:**

Normalising values where necessary



**Reading a new dataset Calorie_df and preparing it for analysis**

In [ ]:
import pandas as pd
url = "https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/Data%20Cleaning%20Data%20Sets/dirtydata.csv?raw=true"
calorie_df=pd.read_csv(url)

calorie_df

Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'     90       112       NaN
19        60  '2020/12/19'    103       123     323.0
20        45  '2020/12/20'     97       125     243.0
21        60  '2020/12/21'    108       131     364.2
22        45           NaN    100       119     282.0
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132       NaN
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0

In the below example, I have used .loc() to locate a record with incorrect value and have assigned a correct value to it.

In [ ]:
#to replace incorrect value for Duration with the right one

calorie_df1=calorie_df
calorie_df1.loc[calorie_df1['Duration'] ==450, 'Duration'] = 45
calorie_df1

Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7         45  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'     90       112       NaN
19        60  '2020/12/19'    103       123     323.0
20        45  '2020/12/20'     97       125     243.0
21        60  '2020/12/21'    108       131     364.2
22        45           NaN    100       119     282.0
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132       NaN
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0

Below, I have used df.info() to check which columns contain null values.

In [ ]:
calorie_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32 entries, 0 to 31
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  32 non-null     int64  
 1   Date      31 non-null     object 
 2   Pulse     32 non-null     int64  
 3   Maxpulse  32 non-null     int64  
 4   Calories  30 non-null     float64
dtypes: float64(1), int64(3), object(1)
memory usage: 1.4+ KB


Below I have replaced null values by for calories column by average value. I have used groupby('Duration') as I wanted to fill null values with average calories for the perticular duration.

In [ ]:
#Fill null values by average value grouped by duration

calorie_df2=calorie_df1
calorie_df2['Calories']=calorie_df2.groupby('Duration')['Calories'].transform(lambda x:x.fillna(x.mean()))
calorie_df2['Calories'] = calorie_df2['Calories'].apply(lambda x: round(x, 1))
calorie_df2

Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7         45  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'     90       112     285.4
19        60  '2020/12/19'    103       123     323.0
20        45  '2020/12/20'     97       125     243.0
21        60  '2020/12/21'    108       131     364.2
22        45           NaN    100       119     282.0
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132     314.5
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0

In the below example,
1. I have used drop_duplicates to remove any duplicate records
2. I have used .fillna to replace null value with a correct value
3. I have used to_datetime function to convert date values to correct datatype.

In [ ]:
calorie_df3 = calorie_df2

calorie_df3=calorie_df3.drop_duplicates() #to remove duplicate record

calorie_df4=calorie_df3
calorie_df4['Date'].fillna("2020-12-22", inplace=True) #to replace null value with correct value

calorie_df4['Date']=pd.to_datetime(calorie_df4['Date']) #to convert values from date column to correct datatype
calorie_df4

/usr/local/lib/python3.7/dist-packages/pandas/core/series.py:4536: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  downcast=downcast,
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7         45 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112     285.4
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
22        45 2020-12-22    100       119     282.0
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132     314.5
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

Below, I have used the google colab module to import the function "files". I have then used 'files.download' to download  the "calorie_data.csv" file to my laptop.



In [ ]:
#to download dataset as csv file

from google.colab import files

calorie_df4.to_csv('calorie_data.csv') 
files.download('calorie_data.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

**Preparing Presidential election voting dataset for analysis**

In [ ]:
import pandas as pd
url = "https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/Data%20Cleaning%20Data%20Sets/presDirty.csv?raw=true"
voting_df=pd.read_csv(url)

voting_df.head(20)

cycle     branch  ...         timestamp simulations
0    2020  President  ...  03/11/2020 00:00       40000
1    2020  President  ...  02/11/2020 23:45       40000
2    2020  President  ...  02/11/2020 23:45       40000
3    2020  President  ...  31/10/2020 22:23       40000
4    2020  President  ...  30/10/2020 21:34       40000
5    2020  President  ...  29/10/2020 22:17       40000
6    2020  President  ...  28/10/2020 21:36       40000
7    2020  President  ...  27/10/2020 21:01       40000
8    2020  President  ...  26/10/2020 20:38       40000
9    2020  President  ...  25/10/2020 20:47       40000
10   2020  President  ...  24/10/2020 22:24       40000
11   2020  President  ...  23/10/2020 22:19       40000
12   2020  President  ...  22/10/2020 19:00       40000
13   2020  President  ...  21/10/2020 20:34       40000
14   2020  President  ...  20/10/2020 21:36       40000
15   2020  President  ...  19/10/2020 21:55       40000
16   2020  President  ...  18/10/2020 21:11       40000
17   2020  President  ...  17/10/2020 19:17       40000
18   2020  President  ...  16/10/2020 21:08       40000
19   2020  President  ...  15/10/2020 20:38       40000

[20 rows x 16 columns]

Using df.info() to check columns with null values

In [ ]:
voting_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 16 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   cycle                       156 non-null    int64  
 1   branch                      156 non-null    object 
 2   model                       156 non-null    object 
 3   modeldate                   156 non-null    object 
 4   candidate_inc               156 non-null    object 
 5   candidate_chal              156 non-null    object 
 6   nat_voteshare_other_hi      156 non-null    float64
 7   national_voteshare_inc_lo   156 non-null    float64
 8   national_voteshare_chal_lo  156 non-null    float64
 9   national_voteshare_3rd_lo   0 non-null      float64
 10  nat_voteshare_other_lo      156 non-null    float64
 11  national_turnout            59 non-null     float64
 12  national_turnout_hi         59 non-null     float64
 13  national_turnout_lo         59 non-

Below, I have used df.drop() and specified names for empty columns in the paranthesis to remove them.

In [ ]:
#to remove empty columns

voting_df1=voting_df
voting_df1.drop(['cycle','branch','model','national_voteshare_3rd_lo'] ,inplace=True, axis=1)
voting_df1

modeldate candidate_inc  ...         timestamp  simulations
0    11/03/2020         Trump  ...  03/11/2020 00:00        40000
1    11/02/2020         Trump  ...  02/11/2020 23:45        40000
2    11/02/2020         Trump  ...  02/11/2020 23:45        40000
3    10/31/2020         Trump  ...  31/10/2020 22:23        40000
4    10/30/2020         Trump  ...  30/10/2020 21:34        40000
..          ...           ...  ...               ...          ...
151  06/05/2020         Trump  ...  10/08/2020 11:40        40000
152  06/04/2020         Trump  ...  10/08/2020 12:03        40000
153  06/03/2020         Trump  ...  10/08/2020 11:24        40000
154  06/02/2020         Trump  ...  10/08/2020 11:34        40000
155  06/01/2020         Trump  ...  10/08/2020 11:53        40000

[156 rows x 12 columns]

I have used df.at to locate a record which has empty value and replaced it with correct value.

In [ ]:
#to replace empty value(space) with correct value

voting_df2=voting_df1
voting_df2.at[60,'modeldate']='09/04/2020'
voting_df2.iloc[50:61]

modeldate candidate_inc  ...         timestamp  simulations
50   9/14/2020         Trump  ...  14/09/2020 23:33        40000
51   9/13/2020         Trump  ...  13/09/2020 19:00        40000
52  09/12/2020         Trump  ...  12/09/2020 19:00        40000
53  09/11/2020         Trump  ...  11/09/2020 19:00        40000
54  09/10/2020         Trump  ...  10/09/2020 21:47        40000
55  09/09/2020         Trump  ...  09/09/2020 23:52        40000
56  09/08/2020         Trump  ...  08/09/2020 22:02        40000
57  09/07/2020         Trump  ...  07/09/2020 21:01        40000
58  09/06/2020         Trump  ...  06/09/2020 21:06        40000
59  09/05/2020         Trump  ...  05/09/2020 22:17        40000
60  09/04/2020         Trump  ...  04/09/2020 22:34        40000

[11 rows x 12 columns]

Again, used dropna() to remove all records with null values.

In [ ]:
#to remove records with any null value
voting_df3=voting_df2.dropna()
voting_df3

modeldate candidate_inc  ...         timestamp  simulations
0   11/03/2020         Trump  ...  03/11/2020 00:00        40000
1   11/02/2020         Trump  ...  02/11/2020 23:45        40000
2   11/02/2020         Trump  ...  02/11/2020 23:45        40000
3   10/31/2020         Trump  ...  31/10/2020 22:23        40000
4   10/30/2020         Trump  ...  30/10/2020 21:34        40000
5   10/29/2020         Trump  ...  29/10/2020 22:17        40000
6   10/28/2020         Trump  ...  28/10/2020 21:36        40000
7   10/27/2020         Trump  ...  27/10/2020 21:01        40000
8   10/26/2020                ...  26/10/2020 20:38        40000
9   10/25/2020         Trump  ...  25/10/2020 20:47        40000
10  10/24/2020         Trump  ...  24/10/2020 22:24        40000
11  10/23/2020         Trump  ...  23/10/2020 22:19        40000
12  10/22/2020         Trump  ...  22/10/2020 19:00        40000
13  10/21/2020         Trump  ...  21/10/2020 20:34        40000
14  10/20/2020         Trump  ...  20/10/2020 21:36        40000
15  10/19/2020         Trump  ...  19/10/2020 21:55        40000
16  10/18/2020         Trump  ...  18/10/2020 21:11        40000
17  10/17/2020         Trump  ...  17/10/2020 19:17        40000
18  10/16/2020         Trump  ...  16/10/2020 21:08        40000
19  10/15/2020                ...  15/10/2020 20:38        40000
20  10/14/2020         Trump  ...  14/10/2020 21:42        40000
21  10/13/2020         Trump  ...  13/10/2020 23:16        40000
22  10/12/2020         Trump  ...  12/10/2020 21:07        40000
23  10/11/2020         Trump  ...  11/10/2020 19:00        40000
24  10/10/2020         Trump  ...  10/10/2020 19:00        40000
25  10/09/2020         Trump  ...  09/10/2020 19:46        40000
26  10/08/2020         Trump  ...  08/10/2020 21:25        40000
27  10/07/2020         Trump  ...  07/10/2020 20:48        40000
28  10/06/2020         Trump  ...  06/10/2020 22:00        40000
29  10/05/2020         Trump  ...  05/10/2020 23:45        40000
30  10/04/2020         Trump  ...  04/10/2020 20:24        40000
31  10/03/2020         Trump  ...  03/10/2020 20:48        40000
32  10/02/2020         Trump  ...  02/10/2020 22:26        40000
33  10/01/2020         Trump  ...  01/10/2020 21:27        40000
34   9/30/2020         Trump  ...  30/09/2020 20:24        40000
35   9/29/2020         Trump  ...  29/09/2020 20:44        40000
36   9/28/2020         Trump  ...  28/09/2020 22:13        40000
37   9/27/2020         Trump  ...  27/09/2020 20:00        40000
38   9/26/2020         Trump  ...  26/09/2020 19:00        40000
39   9/25/2020         Trump  ...  25/09/2020 22:17        40000
40   9/24/2020         Trump  ...  24/09/2020 19:10        40000
41   9/23/2020         Trump  ...  23/09/2020 20:53        40000
42   9/22/2020         Trump  ...  22/09/2020 19:00        40000
43   9/21/2020         Trump  ...  21/09/2020 19:00        40000
44   9/20/2020         Trump  ...  20/09/2020 22:02        40000
45   9/19/2020         Trump  ...  19/09/2020 19:00        40000
46   9/18/2020         Trump  ...  18/09/2020 19:20        40000
47   9/17/2020         Trump  ...  17/09/2020 19:00        40000
48   9/16/2020         Trump  ...  16/09/2020 23:12        40000
49   9/15/2020         Trump  ...  15/09/2020 23:33        40000
50   9/14/2020         Trump  ...  14/09/2020 23:33        40000
51   9/13/2020         Trump  ...  13/09/2020 19:00        40000
52  09/12/2020         Trump  ...  12/09/2020 19:00        40000
53  09/11/2020         Trump  ...  11/09/2020 19:00        40000
54  09/10/2020         Trump  ...  10/09/2020 21:47        40000
55  09/09/2020         Trump  ...  09/09/2020 23:52        40000
56  09/08/2020         Trump  ...  08/09/2020 22:02        40000
57  09/07/2020         Trump  ...  07/09/2020 21:01        40000
58  09/06/2020         Trump  ...  06/09/2020 21:06        40000

[59 rows x 12 columns]

Again,used df.at to locate a record which has empty values and replaced it with correct value.

In [ ]:
#to replace space with correct value

voting_df3.at[8,'candidate_inc']='Trump'
voting_df3.at[19,'candidate_inc']='Trump'
voting_df3

modeldate candidate_inc  ...         timestamp  simulations
0   11/03/2020         Trump  ...  03/11/2020 00:00        40000
1   11/02/2020         Trump  ...  02/11/2020 23:45        40000
2   11/02/2020         Trump  ...  02/11/2020 23:45        40000
3   10/31/2020         Trump  ...  31/10/2020 22:23        40000
4   10/30/2020         Trump  ...  30/10/2020 21:34        40000
5   10/29/2020         Trump  ...  29/10/2020 22:17        40000
6   10/28/2020         Trump  ...  28/10/2020 21:36        40000
7   10/27/2020         Trump  ...  27/10/2020 21:01        40000
8   10/26/2020         Trump  ...  26/10/2020 20:38        40000
9   10/25/2020         Trump  ...  25/10/2020 20:47        40000
10  10/24/2020         Trump  ...  24/10/2020 22:24        40000
11  10/23/2020         Trump  ...  23/10/2020 22:19        40000
12  10/22/2020         Trump  ...  22/10/2020 19:00        40000
13  10/21/2020         Trump  ...  21/10/2020 20:34        40000
14  10/20/2020         Trump  ...  20/10/2020 21:36        40000
15  10/19/2020         Trump  ...  19/10/2020 21:55        40000
16  10/18/2020         Trump  ...  18/10/2020 21:11        40000
17  10/17/2020         Trump  ...  17/10/2020 19:17        40000
18  10/16/2020         Trump  ...  16/10/2020 21:08        40000
19  10/15/2020         Trump  ...  15/10/2020 20:38        40000
20  10/14/2020         Trump  ...  14/10/2020 21:42        40000
21  10/13/2020         Trump  ...  13/10/2020 23:16        40000
22  10/12/2020         Trump  ...  12/10/2020 21:07        40000
23  10/11/2020         Trump  ...  11/10/2020 19:00        40000
24  10/10/2020         Trump  ...  10/10/2020 19:00        40000
25  10/09/2020         Trump  ...  09/10/2020 19:46        40000
26  10/08/2020         Trump  ...  08/10/2020 21:25        40000
27  10/07/2020         Trump  ...  07/10/2020 20:48        40000
28  10/06/2020         Trump  ...  06/10/2020 22:00        40000
29  10/05/2020         Trump  ...  05/10/2020 23:45        40000
30  10/04/2020         Trump  ...  04/10/2020 20:24        40000
31  10/03/2020         Trump  ...  03/10/2020 20:48        40000
32  10/02/2020         Trump  ...  02/10/2020 22:26        40000
33  10/01/2020         Trump  ...  01/10/2020 21:27        40000
34   9/30/2020         Trump  ...  30/09/2020 20:24        40000
35   9/29/2020         Trump  ...  29/09/2020 20:44        40000
36   9/28/2020         Trump  ...  28/09/2020 22:13        40000
37   9/27/2020         Trump  ...  27/09/2020 20:00        40000
38   9/26/2020         Trump  ...  26/09/2020 19:00        40000
39   9/25/2020         Trump  ...  25/09/2020 22:17        40000
40   9/24/2020         Trump  ...  24/09/2020 19:10        40000
41   9/23/2020         Trump  ...  23/09/2020 20:53        40000
42   9/22/2020         Trump  ...  22/09/2020 19:00        40000
43   9/21/2020         Trump  ...  21/09/2020 19:00        40000
44   9/20/2020         Trump  ...  20/09/2020 22:02        40000
45   9/19/2020         Trump  ...  19/09/2020 19:00        40000
46   9/18/2020         Trump  ...  18/09/2020 19:20        40000
47   9/17/2020         Trump  ...  17/09/2020 19:00        40000
48   9/16/2020         Trump  ...  16/09/2020 23:12        40000
49   9/15/2020         Trump  ...  15/09/2020 23:33        40000
50   9/14/2020         Trump  ...  14/09/2020 23:33        40000
51   9/13/2020         Trump  ...  13/09/2020 19:00        40000
52  09/12/2020         Trump  ...  12/09/2020 19:00        40000
53  09/11/2020         Trump  ...  11/09/2020 19:00        40000
54  09/10/2020         Trump  ...  10/09/2020 21:47        40000
55  09/09/2020         Trump  ...  09/09/2020 23:52        40000
56  09/08/2020         Trump  ...  08/09/2020 22:02        40000
57  09/07/2020         Trump  ...  07/09/2020 21:01        40000
58  09/06/2020         Trump  ...  06/09/2020 21:06        40000

[59 rows x 12 columns]

Below, I have used the google colab module to import the function "files". I have then used 'files.download' to download the csv file to my laptop which is clean and ready for analysis.



In [ ]:
#to download dataset as csv file

from google.colab import files
voting_df3.to_csv('voting_data.csv') 
files.download('voting_data.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

**Prepare dataset for normalisation**

Below I have read Covid vaccination data from the by_country sheet in the given Excel file. 

In [ ]:
# 1 Read the data
import pandas as pd
url="https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true"
covid=pd.read_excel(url,sheet_name="by_country")
covid.head()

country  ...                                     source_website
0  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
1  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
2  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
3  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
4  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...

[5 rows x 15 columns]

In the below cell, I have used df.info() to find out which columns have null values.

In [ ]:
#2 Use .info() to find out which columns have missing values
covid.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14994 entries, 0 to 14993
Data columns (total 15 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   country                              14994 non-null  object        
 1   iso_code                             14994 non-null  object        
 2   date                                 14994 non-null  datetime64[ns]
 3   total_vaccinations                   9011 non-null   float64       
 4   people_vaccinated                    8370 non-null   float64       
 5   people_fully_vaccinated              6158 non-null   float64       
 6   daily_vaccinations_raw               7575 non-null   float64       
 7   daily_vaccinations                   14796 non-null  float64       
 8   total_vaccinations_per_hundred       9011 non-null   float64       
 9   people_vaccinated_per_hundred        8370 non-null   float64       
 10  people_ful

Below, I have used dropna() with the parameter 'subset' to remove rows with missing values for the perticular column.

In [ ]:
#3 Remove all rows with missing data in the total_vaccination column
covid.dropna(subset=["total_vaccinations"])

country  ...                                     source_website
0      Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
6      Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
22     Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
44     Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
59     Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
...            ...  ...                                                ...
14989     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...
14990     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...
14991     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...
14992     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...
14993     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...

[9011 rows x 15 columns]

Below code finds the median vaccinations per hundred. .median() function is used.

In [ ]:
#4 Find the median vaccinations per hundred
import pandas as pd
url="https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true"
covid=pd.read_excel(url,sheet_name="by_country")
covid['total_vaccinations_per_hundred'].median()

I wanted to display the mean people vaccinated per hundred for each country in descending order. So, I created a new dataframe with only 'people vaccinated per hundred' column. Used dropna on this column to remove records with null values. Then used groupby function with subset 'country' and used .mean() to find out average for each country. Also, used ascending=False parameter to display the mean people vaccinated per hundred for each country in descending order.

In [ ]:
#5 display the mean people vaccinated per hundred for each country in descending order
new_df=covid.dropna(subset=['people_vaccinated_per_hundred']) # new df with column people_vaccinated_per_hundred with null records removed
sorted=new_df[['people_vaccinated_per_hundred','country']].groupby(['country']).mean().sort_values(['people_vaccinated_per_hundred'],ascending = False)
sorted

people_vaccinated_per_hundred
country                                                    
Gibraltar                                         64.975699
Bhutan                                            55.961892
Falkland Islands                                  51.063333
Saint Helena                                      44.880000
Seychelles                                        44.005686
...                                                     ...
Syria                                              0.005000
South Sudan                                        0.005000
Niger                                              0.005000
Cameroon                                           0.000000
Democratic Republic of Congo                       0.000000

[189 rows x 1 columns]

Below, I wanted to find the range of total_vaccinations across the dataframe. So, I created a new dataframe with a column total_vaccinations from original dataframe. Used dropna to drop any records with null values. Then found out range by subtracting minimum of total_vaccinations from maximum of total_vaccinations.

In [ ]:
#6 find the range of total_vaccinations across the dataframe
covid1=covid.dropna(subset=["total_vaccinations"])
range=covid1['total_vaccinations'].max()-covid1['total_vaccinations'].min()
range

275338000.0

**Normalise daily vaccinations**


First, I found out the median daily vaccinations per million using .median().
Then wrote a function to normalise daily vaccinations per million, where values greater than or equal to median = 1 and values less than median = 0

In [ ]:
#median of daily_vaccinations_per_million
medianV=covid['daily_vaccinations_per_million'].median()
print(medianV)

#normalise daily_vaccinations_per_million
def normalise_d_v_m(df):
  if df['daily_vaccinations_per_million']< medianV:
    return 0
  else:
    return 1


covid['daily_vaccinations_per_million']=covid.apply(normalise_d_v_m,axis=1)
covid['daily_vaccinations_per_million']
    


1475.0


0        1
1        0
2        0
3        0
4        0
        ..
14989    0
14990    0
14991    0
14992    0
14993    0
Name: daily_vaccinations_per_million, Length: 14994, dtype: int64

**Normalising total vaccinations**

I wanted to find the minimum total vaccinations for the United Kingdom. So, first I created a new dataframe with only two columns, country and minimum total vaccinations for each country. I used .groupby('country').min() to find out minimum total vaccinations for each country. Then, I used df.loc with subset 'united kingdom' to find out min vaccinations for United Kingdom. I used int() function to round this min vaccinations for UK value.
Then, I wrote a function to normalise total_vaccinations column so that all values less than the UK's min are 0 and all values greater than or equal to the UK's min are coded as 1

In [ ]:
#min total vaccinations for uk
import pandas as pd
url="https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true"
covid=pd.read_excel(url,sheet_name="by_country")
covid.head()
covid1=covid[['total_vaccinations','country']].groupby('country').min() #minimum total_vaccinations for each country

min=covid1.loc["United Kingdom"]['total_vaccinations'] #minimum total_vaccinations for UK

min=int(min) #round value for min total uk vaccinations
min

#function to normalise total_vaccinations column
def normalise_vaccinations(df):
  if df['total_vaccinations']<min:
    return 0
  else:
    return 1

covid.dropna(subset=['total_vaccinations'])
covid['total_vaccinations']=covid.apply(normalise_vaccinations,axis=1)
covid['total_vaccinations']


0        0
1        1
2        1
3        1
4        1
        ..
14989    0
14990    0
14991    0
14992    0
14993    0
Name: total_vaccinations, Length: 14994, dtype: int64

I wanted to display the countries for which total vaccinated is at the same rate or more than the UK. I created new dataframe with sum of total_vaccinations for each country. Used groupby('country') and .sum().
Then used .loc function to find total vaccinations for UK.
Then created a list country to store countries for which total vaccinated is at the same rate or more than the UK.

In [ ]:
#Display the countries for which total vaccinated is at the same rate or more than the UK
import pandas as pd
url="https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true"
covid=pd.read_excel(url,sheet_name="by_country")
covid1=covid[['total_vaccinations','country']].groupby('country').sum() #sum of total_vaccinations for each country

total=covid1.loc["United Kingdom"]['total_vaccinations'] #find out total vaccinations for UK
print(total)

#find countries for which total vaccinated is at the same rate or more than the UK.
country=[]
for i in covid1.index:
  if covid1.loc[i]['total_vaccinations']>=total and i !='United Kingdom': 
    country.append(i)

country


2748464937.0


['China', 'India', 'United States']

**create new series of total vaccinations for each manufacturer**

Reading the manufacturer data

In [ ]:
#read manufacturer data
import pandas as pd
url="https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true"
covidM=pd.read_excel(url,sheet_name="by_manufacturer")
covidM.head()

location       date          vaccine  total_vaccinations
0    Chile 2020-12-24  Pfizer/BioNTech                 420
1    Chile 2020-12-25  Pfizer/BioNTech                5198
2    Chile 2020-12-26  Pfizer/BioNTech                8338
3    Chile 2020-12-27  Pfizer/BioNTech                8649
4    Chile 2020-12-28  Pfizer/BioNTech                8649

Used groupby('vaccine').sum() to find out total vaccinations for each manufacturer

In [ ]:
#2 find the sum of total vaccinations for each manufacturer
covidM[['total_vaccinations','vaccine']].groupby('vaccine').sum()


total_vaccinations
vaccine                               
Johnson&Johnson              264839828
Moderna                     5548036383
Oxford/AstraZeneca           539433203
Pfizer/BioNTech             8690461304
Sinovac                      604660293

created a new column that has the total vaccinations as a percentage of the overall sum of total vaccinations

In [ ]:
#3 create a new column that has the total vaccinations as a percentage of the overall sum of total vaccinations
covidM['total_vaccinations_percentage']=(covidM['total_vaccinations']/covidM['total_vaccinations'].sum()) * 100
covidM.head(10)


location       date  ... total_vaccinations  total_vaccinations_percentage
0    Chile 2020-12-24  ...                420                       0.000003
1    Chile 2020-12-25  ...               5198                       0.000033
2    Chile 2020-12-26  ...               8338                       0.000053
3    Chile 2020-12-27  ...               8649                       0.000055
4    Chile 2020-12-28  ...               8649                       0.000055
5    Chile 2020-12-29  ...               8649                       0.000055
6    Chile 2020-12-30  ...               8649                       0.000055
7    Chile 2020-12-31  ...               8649                       0.000055
8    Chile 2021-01-01  ...               8649                       0.000055
9    Chile 2021-01-02  ...               8649                       0.000055

[10 rows x 5 columns]

Below code is to find the median percentage

In [ ]:
#4 find the median percentage
median_percentage=covidM['total_vaccinations_percentage'].median()
median_percentage

0.0011110194374896931

Created duplicate column of total_vaccinations_percentage column

In [ ]:
#5 create a new column called 'normalised_percentages' which duplicates the percentages column
covidM['normalised_percentage']=covidM['total_vaccinations_percentage']
covidM.head(10)

location       date  ... total_vaccinations_percentage  normalised_percentage
0    Chile 2020-12-24  ...                      0.000003               0.000003
1    Chile 2020-12-25  ...                      0.000033               0.000033
2    Chile 2020-12-26  ...                      0.000053               0.000053
3    Chile 2020-12-27  ...                      0.000055               0.000055
4    Chile 2020-12-28  ...                      0.000055               0.000055
5    Chile 2020-12-29  ...                      0.000055               0.000055
6    Chile 2020-12-30  ...                      0.000055               0.000055
7    Chile 2020-12-31  ...                      0.000055               0.000055
8    Chile 2021-01-01  ...                      0.000055               0.000055
9    Chile 2021-01-02  ...                      0.000055               0.000055

[10 rows x 6 columns]

Wrote a function to normalise the normalised_percentages column so that any values greater than or equal to the median percentage = 1 and any lesser than = 0

In [ ]:
#6 normalise the normalised_percentages column so that any values greater than or equal to the median percentage = 1 and any lesser than = 0
def normalise_percentage(df):
  if df['total_vaccinations_percentage']>=median_percentage:
    return 1
  else:
    return 0


covidM['normalised_percentage']=covidM.apply(normalise_percentage,axis=1)
covidM

location  ... normalised_percentage
0             Chile  ...                     0
1             Chile  ...                     0
2             Chile  ...                     0
3             Chile  ...                     0
4             Chile  ...                     0
...             ...  ...                   ...
3291  United States  ...                     1
3292  United States  ...                     1
3293  United States  ...                     1
3294  United States  ...                     1
3295  United States  ...                     1

[3296 rows x 6 columns]

**Skills covered:**

Correcting the format of dates where necessary



Below, I have written a function which will take a date and display it in parts as specified. I used datetime function from datetime library.I used .year,.month to display year,month of the given date. Used strftime function to display date in differnt formats.

In [ ]:
#display parts of date/time
from datetime import datetime
def display_date_parts(date_to_check):
  print("Given date:",date_to_check)
  print("Year:",date_to_check.year)
  print("Month in number:",date_to_check.month)
  print("Month name:",date_to_check.strftime("%B"))
  print("Week number of the year:",date_to_check.strftime("%U"))
  print("Weekday of the week in number:",date_to_check.strftime("%w"))
  print("Day of the year:",date_to_check.strftime("%j"))
  print("Day of the month:",date_to_check.strftime("%d"))
  print("Day of the week as name:",date_to_check.strftime("%A"))






check_date=datetime(1998,12,23)
display_date_parts(check_date)

  

Given date: 1998-12-23 00:00:00
Year: 1998
Month in number: 12
Month name: December
Week number of the year: 51
Weekday of the week in number: 3
Day of the year: 357
Day of the month: 23
Day of the week as name: Wednesday


Below, I have written a function that takes a date and generate the dates of the following 6 days. To generate the days, I have used the function datetime.timedelta()

In [ ]:
#Print dates of the 6 days followed by given date
from datetime import timedelta,datetime
def generate_nextSix_days(date_passed):
  #ini_time_for_now + timedelta(days = 730)
  for i in range(1,7):
    new_date=date_passed+timedelta(days=i)
    print(new_date)
    print("Day of the week:",new_date.strftime("%w"),",Month:",new_date.strftime("%B"),",Year:",new_date.year)


date=datetime(2020,12,28)
generate_nextSix_days(date)


2020-12-29 00:00:00
Day of the week: 2 ,Month: December ,Year: 2020
2020-12-30 00:00:00
Day of the week: 3 ,Month: December ,Year: 2020
2020-12-31 00:00:00
Day of the week: 4 ,Month: December ,Year: 2020
2021-01-01 00:00:00
Day of the week: 5 ,Month: January ,Year: 2021
2021-01-02 00:00:00
Day of the week: 6 ,Month: January ,Year: 2021
2021-01-03 00:00:00
Day of the week: 0 ,Month: January ,Year: 2021


**Skills Covered**



1.  Resetting data values
2.  Joining data from 2 different sources
3.  Merging and joining data





**Data Wrangling**

Below I have written a function which
reads the three sheets into the dataframes county_level_df, state_level_df and income_new_df
cleans up the Population column of county_level_df so that it has the name Pop
cleans up the Pop column of state_level_df so that it contains values that are whole numbers of 1000s
combines county_level_df and state_level_df
appends income_new_df to the combined dataframe
SORTs the final dataframe on Income in descending order and
return the final dataframe

In [1]:
import pandas as pd

def get_income_df(url):
  county_level_df=pd.read_excel(url,sheet_name="county-level")

  state_level_df=pd.read_excel(url,sheet_name="state-level")

  income_new_df=pd.read_excel(url,sheet_name="income")

  county_level_df['Pop']=county_level_df['Population']  #to change name of column
  county_level_df.drop(['Population'],axis=1)

  state_level_df['Pop']=round(state_level_df['Pop']/1000) #to convert values to whole number
  state_level_df['Pop']=state_level_df['Pop'].astype(int)

  combined_df=pd.concat([county_level_df,state_level_df], join='inner', ignore_index=True) #to combine two dataframes
   
  income_df_v=income_new_df.append(combined_df,ignore_index=True) #to add new dataframe to combined dataframe
 

  income_df_v4=income_df_v.sort_values(['Income'],ascending=False) #to sort dataframe by Income in descending order
  
  return income_df_v4



url="https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/Income-Data.xlsx?raw=true"
income_df=get_income_df(url)
income_df



State  Age  Income  Pop
25    MD   22      81   89
32    NY   34      78   18
4     NY   29      78   11
24    MD   31      78   34
8     CA   29      78   10
17    TX   40      75    5
21    MD   29      75   64
26    MD   25      73   21
20    MD   25      73   98
29    MD   29      69   10
19    MD   23      69    5
0     TX   32      69   70
2     IN   32      69   23
9     NY   34      67    7
31    CA   35      67   30
6     MD   34      67   81
16    TX   35      66   82
23    MD   25      66   24
13    TX   36      65   54
22    MD   24      65   36
10    TX   34      65   72
27    MD   30      62   21
15    TX   25      62   28
5     TX   35      54   27
3     CA   35      54   54
28    TX   32      54   24
14    TX   42      53   11
12    TX   23      46   25
11    TX   42      45   33
7     IN   41      35   24
1     MD   41      35   33
30    IN   41      35    5
18    MD   27      22   61

Below I have run the function get_income_df() and store the result in a new dataframe called income_df.ANd then grouped income_df by State and calculated some statistics on the groups.

In [3]:
#to get avearge of income and population for each state

income_df=get_income_df(url)
income_df[['State','Income','Pop']].groupby(['State']).mean()

Income        Pop
State                      
CA     66.333333  31.333333
IN     46.333333  17.333333
MD     64.230769  44.384615
NY     74.333333  12.000000
TX     59.454545  39.181818

In [4]:
#to get lowest of income and population for each state

income_df=get_income_df(url)
income_df[['State','Income','Pop']].groupby(['State']).min()

Income  Pop
State             
CA         54   10
IN         35    5
MD         22    5
NY         67    7
TX         45    5

In [5]:
#to get highest of income and population for each state

income_df=get_income_df(url)
income_df[['State','Income','Pop']].groupby(['State']).max()

Income  Pop
State             
CA         78   54
IN         69   24
MD         81   98
NY         78   18
TX         75   82

In [6]:
#to get statistics of Income, Age, Population for each state

income_df=get_income_df(url)
income_df.groupby('State').agg(
    {
       "Age" :['min','max','mean','std','median'],
       "Income":['min','max','mean','std','median'],
       "Pop" :['min','max','mean','std','median']
    }
)

Age                                  ... Pop                                 
      min max       mean       std median  ... min max       mean        std median
State                                      ...                                     
CA     29  35  33.000000  3.464102     35  ...  10  54  31.333333  22.030282     30
IN     32  41  38.000000  5.196152     41  ...   5  24  17.333333  10.692677     23
MD     22  41  28.076923  5.203549     27  ...   5  98  44.384615  30.845147     34
NY     29  34  32.333333  2.886751     34  ...   7  18  12.000000   5.567764     11
TX     23  42  34.181818  6.161464     35  ...   5  82  39.181818  26.049254     28

[5 rows x 15 columns]

These statstics shows, Maryland has lowest average age and highest average population.Although New York has lowest average population, it's average income is highest.Average age for Texas, New York and California is close by but still there is a big differance between average income and average population for these 3 states.

Indiana and Maryland have the same maximum age but quiet different minimum age. For population these two states have same minimum population but a big differance in maximum population.

